WIP: Virtual readbacks

hrydgard opened the pull request:
As outlined in #11528 

Fixes performance in Digimon Adventures and lets us remove the old MotoGP hack.

In addition there are multiple other games this could help but we'll add them later.

hrydgard committed:
Add compat flag to allow virtual framebuffer readbacks (auto-create fb from readback destination). Does not yet work outside VRAM but should fix Digimon Adventure.

hrydgard commented on GPU/Common/FramebufferCommon.cpp:
```diff
@@ -1419,6 +1420,38 @@ void FramebufferManagerCommon::FindTransferFramebuffers(VirtualFramebuffer *&dst
+		vfbs_.push_back(vfb);
+		dstBuffer = vfb;
+	}
+	dstBuffer->last_frame_used = gpuStats.numFlips;
```
At that point, dstBuffer is guaranteed not to be old. Plus, these targets that we don't otherwise render to but texture from do need their last_frame_used updated, otherwise they'll quickly be decimated.

unknownbrackets replied:
When `BlockTransferAllowCreateFB` is false, if`dstBuffer` cannot be null - why do we even need the `!dstBuffer &&` check above, then?

-[Unknown]

hrydgard replied:
Um.. right. Oops, sorry was too hasty :)

hrydgard commented:
Reportedly the speedup is massive in Digimon Adventures and the toon borders are sharp, with this.

So seems promising.

I really want to rethink and unify textures, framebuffers and CLUTs long term, but some ideas I have would require at least ES 3.0 level features and preferably higher. I really wish we could just forget about ES 2.0 and DX9 but we can't. So we're probably stuck with our current pile of hacks for some time longer, and this is just an addition to it. I will improve this further before merging of course, at least to fix the MotoGP case too so we can delete one hack while adding this one.

hrydgard commented on GPU/Common/FramebufferCommon.cpp:
```diff
@@ -1419,6 +1420,38 @@ void FramebufferManagerCommon::FindTransferFramebuffers(VirtualFramebuffer *&dst
+		vfb->fb_address = dstBasePtr;  // NOTE - not necessarily in VRAM!
+		vfb->fb_stride = dstStride;
+		vfb->z_address = 0;
+		vfb->z_stride =0;
```
Yeah, that's a good point. This can be quite dangerous otherwise...

hrydgard committed:
Support virtual readbacks for detected-memcpy framebuffer copies, delete MotoGP hack.

hrydgard commented:
Rebased on #11553 and then master, added support in the detected-memcpy path as well and killed the old MotoGP hack.

This can actually also be used to avoid the sun flare readbacks in Burnout, the sun flare is working but it feels slightly different somehow, maybe because higher resolution. Not sure if I want to enable that.

hrydgard committed:
Take care of most cases of render-to-virtual-readback-buffer, I think.

unknownbrackets commented:
I wonder if we want to apply "lower resolution for effects" to this or anything...

-[Unknown]

hrydgard commented:
Hm, that's an interesting point.. Definitely not for Digimon Adventure outlines, but maybe for Burnout's sun.

hrydgard commented:
I know, I didn't enable this for Burnout yet.
